pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -99,7 +106,83 @@ func (r *MachineSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
 }
 
 // Reconcile reconciles CAPI and MAPI machines for their respective namespaces.
+//
+//nolint:funlen
 func (r *MachineSyncReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", req.Namespace, "name", req.Name)
+
+	logger.V(1).Info("Reconciling machine")
+	defer logger.V(1).Info("Finished reconciling machine")
+
+	var mapiMachineNotFound, capiMachineNotFound bool
+
+	// Get the MAPI Machine.
+	mapiMachine := &machinev1beta1.Machine{}
+	mapiNamespacedName := client.ObjectKey{
+		Namespace: r.MAPINamespace,
+		Name:      req.Name,
+	}
+
+	if err := r.Get(ctx, mapiNamespacedName, mapiMachine); apierrors.IsNotFound(err) {
+		logger.Info("MAPI Machine not found")
+
+		mapiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get MAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("failed to get MAPI machine: %w", err)
+	}
+
+	// Get the corresponding CAPI Machine.
+	capiMachine := &capiv1beta1.Machine{}
+	capiNamespacedName := client.ObjectKey{
+		Namespace: r.CAPINamespace,
+		Name:      req.Name,
+	}
+
+	if err := r.Get(ctx, capiNamespacedName, capiMachine); apierrors.IsNotFound(err) {
+		logger.Info("CAPI Machine not found")
+
+		capiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get CAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("failed to get CAPI machine:: %w", err)
+	}
+
+	if mapiMachineNotFound && capiMachineNotFound {
+		logger.Info("CAPI and MAPI machines not found, nothing to do")
+		return ctrl.Result{}, nil
+	}
+
+	// We mirror if the CAPI machine is owned by a MachineSet which has a MAPI
+	// counterpart. This is because we want to be able to migrate in both directions.
+	if mapiMachineNotFound {
+		if shouldReconcile, err := r.shouldMirrorCAPIMachineToMAPIMachine(ctx, logger, capiMachine); err != nil {
+			return ctrl.Result{}, err
+		} else if shouldReconcile {
+			return r.reconcileCAPIMachinetoMAPIMachine(ctx, capiMachine, mapiMachine)
+		}
+	}
+
+	switch mapiMachine.Spec.AuthoritativeAPI {
```

Comment:
Ah yep this is me misremembering, it should be `Status`! 

I've had a read of the enhancement doc again, I think that this should be handled by a separate controller all together? The `migration-controller`

```
To allow the migration between Machine API and Cluster API resources and vice versa, we will implement a new controller machine-api-migration-controller. This controller will be responsible for co-ordinating the hand over of authority between the Machine API and Cluster API controllers.

The controller will observe the spec and status authoritative API fields and look for a difference between them. A difference indicating the cluster admin requested a migration between the two APIs.

When the spec and status differ, it will first check if the Synchronized condition is set to True, if it is not, this indicates an unmigratable resource and therefore should take no action.

If the condition allows, the controller will set the status.authoritativeAPI to the Migrating value and wait for the old authoritative resource to set the Paused condition to True.

Once the Paused condition is observed, the controller will move the status to match the spec allowing the new authoritative API to take over.

This is the controller that will be responsible for pausing and un-pausing Cluster API resources as part of the handover.
```


I'm not following this: 
>but, if the status is migrating, we need to know where we are coming from, so that we can complete the migration process 

If we're going MAPI -> CAPI, we'll just observe the change of `authoritativeAPI` to be CAPI and start syncing the other way.. or am I missing something?

Reply:
No you're right, we split this out to be a separate controller, clever us! I think you just need to observe status here and handle each of the three cases appropriately. Migrating would just be a no-op :)

Reply:
Sweet - have updated to switch on `Status`